tortoise/tortoise.go

```diff
@@ -695,28 +626,43 @@ func (t *turtle) onHareOutput(lid types.LayerID, bid types.BlockID) {
 	if exists && previous == bid {
 		return
 	}
-	if lid.Before(t.processed) && t.mode.isVerifying() && withinDistance(t.Config.Hdist, lid, t.last) {
+	if lid.Before(t.processed) && !t.isFull && withinDistance(t.Config.Hdist, lid, t.last) {
 		t.logger.With().Info("local opinion changed within hdist",
 			lid,
 			log.Stringer("verified", t.verified),
 			log.Stringer("previous", previous),
 			log.Stringer("new", bid),
 		)
-		t.verifying.resetWeights()
-
-		// if local opinion within hdist was changed about the layer
-		// that was already verified we need to revert that
-		t.verified = minLayer(t.verified, lid)
-		for target := t.verified; !target.After(t.processed); target = target.Add(1) {
-			// TODO(dshulyak) this condition can be removed together with genesis ballot
-			if target.GetEpoch().IsGenesis() {
-				continue
-			}
+
+		t.verifying.resetWeights(lid)
+		for target := lid.Add(1); !target.After(t.processed); target = target.Add(1) {
 			t.verifying.countVotes(t.logger, t.layer(target).ballots)
 		}
 	}
 }
 
+func (t *turtle) onAtx(atx *types.ActivationTxHeader) {
+	epoch, exist := t.epochs[atx.TargetEpoch()]
+	if !exist {
+		epoch = &epochInfo{atxs: map[types.ATXID]uint64{}}
+		t.epochs[atx.TargetEpoch()] = epoch
+	}
+	if _, exist := epoch.atxs[atx.ID]; !exist {
+		t.logger.With().Debug("on atx",
+			log.Stringer("id", atx.ID),
+			log.Uint32("epoch", uint32(atx.TargetEpoch())),
+			log.Uint64("weight", atx.GetWeight()),
+		)
+		epoch.atxs[atx.ID] = atx.GetWeight()
+		epoch.weight += atx.GetWeight()
+		if atx.TargetEpoch() == t.last.GetEpoch() {
```

Comment:
what happens when atx.TargetEpoch < t.last.GetEpoch()? in that case, shouldn't all ballots that references these ATXs added to uncounted good weight for their layers?

Reply:
on atx is always called before ballot that reference it is added to the state, after on atx is called - expected weight and local threshold will change. good uncounted weight will be updated after ballot that referenced this atx is counted.

good uncounted and uncounted would be distinct values. uncounted = expected weight - (total - good uncounted).

Reply:
oh ok. so when processing ballot, if a ballot's layer is already processed, then the ballot is counted towards good uncounted weight.